### PR TITLE
Merge mappings for composable index templates

### PR DESCRIPTION
**Pull request by jtibshirani**
This PR implements recursive mapping merging for composable index templates.

When creating an index, we perform the following:
* Add each component template mapping in order, merging each one in after the
last.
* Merge in the index template mappings (if present).
* Merge in the mappings on the index request itself (if present).

Some principles:
* All 'structural' changes are disallowed (but everything else is fine). An
object mapper can never be changed between `type: object` and `type: nested`. A
field mapper can never be changed to an object mapper, and vice versa.
* Generally, each section is merged recursively. This includes `object`
mappings, as well as root options like `dynamic_templates` and `meta`. Once we
reach 'leaf components' like field definitions, they always overwrite an
existing one instead of being merged.

The PR is pretty large, but each commit can be reviewed individually.

Relates to #53101.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java:112]: The fact that we fix redundant includes makes merging these options more tricky. Previously, mapping merges couldn't make any changes to the `include_in_root` or `include_in_parent` values. Now they can overwrite existing values, so we could have the following situation:
* A new mapper changes `include_in_root` and `include_in_parent` to `true` for some nested object.
* After merging it into the existing one, say the merged mapper now has redundant includes.
* We never detect and fix the redundancy because this method is only called in `build`, not when merging mappers.

I'm wondering if there's a lightweight way to fix this, or some way we can simplify?

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:248]: I think this docstring is incorrect now

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:263]: It's a bit strange to call this "changes" since we're retrieving the mapping for a single field in this method?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:882]: I think I'd rather keep this guard than throw a null pointer exception a couple of lines down, is there a reason you removed it?

Comment (dakrone) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:882]: or at the least, convert this to a real exception rather than an NPE if you want to get rid of it

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:390]: I think this block can be folded into the same check a few lines above rather than having two blocks using the same `if` statement

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:90]: Super minor nit, but most of our error messages are all lowercase and without ending punctuation and I think that's the precedent we want to stick with (applies to the errors below also)
```suggestion
                    throw new IllegalArgumentException("cannot change object mapping from nested to non-nested");
```

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:94]:
```suggestion
                    throw new IllegalArgumentException("cannot change object mapping from non-nested to nested");
```

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:107]:
```suggestion
                    throw new MapperException("the [include_in_parent] parameter can't be updated on a nested object mapping");
```

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:110]:
```suggestion
                    throw new MapperException("the [include_in_root] parameter can't be updated on a nested object mapping");
```

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:482]:
```suggestion
            throw new IllegalArgumentException("can't merge a non object mapping [" + mergeWith.name() + "] with an object mapping");
```

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:502]:
```suggestion
            throw new MapperException("the [enabled] parameter can't be updated for the object mapping [" + name() + "]");
```

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:518]:
```suggestion
                    throw new IllegalArgumentException("can't merge a non object mapping [" +
                        mergeWithMapper.name() + "] with an object mapping");
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/JobManager.java
Comment (dakrone) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/JobManager.java:279]: Yikes that we are doing regex matches on exception messages here :( it feels like there should be a better way, but totally out of scope for this PR

**Review by dakrone — CHANGES_REQUESTED** [review]
Thanks for working on this Julie! I left a couple of comments with (small!) substance and then a bunch of minor nits. I am not an expert in our mapping code so it would be good to have a second opinion also

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:882]: We assert right above that `template != null`, so I thought it'd be safe to simplify this.

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:390]: 👍

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:263]: 👍

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:90]: Oops, I have been adding inconsistent error messages for quite some time now!

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/JobManager.java
Comment (jtibshirani) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/JobManager.java:279]: +1, this was painful to track down and fix.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:882]: I think even with the assertion I'd rather have a helpful message for production clusters than a `NullPointerException` (since assertions are off by default except in tests)

Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:882]: I'll just revert this, I don't feel strongly about it!

File: server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java:112]: Slightly related but we were reluctant to deprecate/remove these options because Kibana doesn't support nested fields. This is something that was stalled for some time but I think we should revisit #12461 now that [nested fields are supported](https://github.com/elastic/kibana/issues/1084#issuecomment-585178079) in Kibana. I'll comment on the issue but I wanted to mention here since it changes the perspective a bit.

> I'm wondering if there's a lightweight way to fix this, or some way we can simplify?

Is there a reason not to call `fixRedundantIncludes` after merging mappers ? That seems like the easiest solution ?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:496]: After chatting with the core-features team, I now realize that we should apply `request.mappings()` *after* this datastream mapping. I'll fix this.

Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:496]: Actually @dakrone is that right? If we applied `request.mappings()` after the datastream timestamp mapping, it could result in an index with an inconsistent timestamp mapping. So perhaps we always need to apply the datastream timestamp mapping last.

File: server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java:112]: So `fixRedundantIncludes` currently works on mapper builders, not mappers themselves. I'll see if I can change this and maybe only call the method once post-mapping merge ?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:496]: I think the `request.mappings()` part last, but we need to validate that the composite mappings are valid for a data stream, that way a user could override part of the mappings (say for instance a custom format for the date), but couldn't do anything bad (like change it to a `keyword` for instance)

**Review by jimczi — APPROVED** [review]
The changes looks great. I like how it is integrated into the mapping merge, it makes the logic very easy to follow. I left one question and it would be nice to have a single place where the full merging logic is described.

**Review by dakrone — APPROVED** [review]
LGTM, thanks Julie!

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:496]: I'm going to keep it as-is for this PR, since it just tries to replicate the existing behavior. We can discuss in a follow-up if we want to change the order.

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:511]: As you mentioned, I think this is general question. Even without this PR you can define an object mapper like this:
```
"object": {
  "type": "object",
  "enabled": false,
  "properties": {
    "field": {
      "type": "keyword"
    }
  }
}
```

For this PR I plan to stick with the existing behavior, but we could have a separate discussion on whether we want to support this flexibility.

**Comment by jtibshirani** [comment]
@elasticmachine elasticsearch-ci/1

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:510]: Agreed, I opened https://github.com/elastic/elasticsearch/issues/58702 to discuss this specifically.

Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:510]: 👍  thanks!

Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:510]: Great thanks!

**Pull request merged**
